### PR TITLE
[ADD] mail_mandrill addon

### PR DESCRIPTION
**Pull request by antespi**
This module logs Mandrill email messages events.
# Configuration

To configure this module, you need to:
- Define an STMP server in Odoo at Settings > Technical > Email > Outgoing Mail Servers
  using Mandrill credentials from Mandrill settings panel (Settings > SMTP & API Credentials)
- Define a webhook in Mandrill settings panel (Settings > Webhooks) with
  several triggers (Message Is Sent, Message Is Delayed, ...) and 'Post to URL'
  like `https://your_odoodomain.com/mandrill/event`
- Copy Webhook key and paste in your Odoo configuration file, in `[options]`
  section, using `mandrill_webhook_key` variable. This is optional, but recommended because it is
  used to validate Mandrill POST requests
# Usage

When any email message is sent via Mandrill SMTP server, Odoo will add
some metadata to email (Odoo DB, Odoo Model and Odoo Model record ID) using an
special SMTP header (X-MC-Metadata). More info at [Mandrill doc: Using Custom Message Metadata](https://mandrill.zendesk.com/hc/en-us/articles/205582417-Using-Custom-Message-Metadata)

Then when an event occurs related with that message (sent, open, click, bounce, ...)
Mandrill will trigger webhook configured and Odoo will log the message and the event.

In `Setting > Technical > Email > Mandrill emails` you can see all messages sent
using Mandrill. When clicking in one of them you'll see message details and events
related with it

In `Setting > Technical > Email > Mandrill events` you can see all Mandrill events
received

### CODE REVIEW

**Comment by antespi** [comment]
Tests pending

**Comment by antespi** [comment]
Sorry @hbrunn, what do you expect to finish in this addon?

**Comment by hbrunn** [comment]
a green travis and coverage >= 0

**Comment by antespi** [comment]
OK, travis is green right now, promising :smile: 
Now is time of testing

**Comment by rafaelbn** [comment]
We are using this module in all our customers, in production environment :+1: tested!
Please remove label work in provess add new review, thanks!

File: mail_mandrill/security/ir.model.access.csv [hunk 1]
```diff
@@ -0,0 +1,3 @@
+"id","name","model_id:id","group_id:id","perm_read","perm_write","perm_create","perm_unlink"
+"access_mail_mandrill_message_group_system","mail_mandrill_message group_system","model_mail_mandrill_message","base.group_system",1,1,1,1
+"access_mail_mandrill_event_group_system","mail_mandrill_event group_system","model_mail_mandrill_event","base.group_system",1,1,1,1
```
Comment (houssine78) [inline mail_mandrill/security/ir.model.access.csv:5]: it would be better to have a dedicated group for this instead of the admin group

**Comment by hbrunn** [comment]
@antespi could you add tests? then it's fit for merging I think

**Comment by antespi** [comment]
@hbrunn Test added and some bugfixes btw :smile:

**Comment by antespi** [comment]
But coveralls is down... coveralls is a misunderstood guy ;)

**Comment by hbrunn** [comment]
@antespi cool! do you think there's something to do with the tests to increase coverage? If there are parts that can't be tested sensibly (need a service account, only for debugging, things like that), use `# pragma: no cover` on the appropriate line(s) or branch(es)

**Comment by hbrunn** [comment]
but on the other hand, coveralls also has the habit to lie... :+1:

**Comment by antespi** [comment]
Yes, that's right. We do not need a service account, but I only added unit test because test all flow is complex

**Comment by antespi** [comment]
I did not test:
- Controller: https://github.com/OCA/social/pull/18/files#diff-a64bc9d4e8a5d7bacc136f6fcb28af79R20
- Send overwrite: https://github.com/OCA/social/pull/18/files#diff-e89dd92ee9127b67461dae6f961eb9a0R33
- Process reject event: https://github.com/OCA/social/pull/18/files#diff-494af699617ae3d2ed2095dabc1e8d6dR166
- Main event process: https://github.com/OCA/social/pull/18/files#diff-3875d1f959ae5a4d2d68dbeef15d52f9R90

How can I add those pragmas

**Comment by hbrunn** [comment]
the pragmas can be added just in your code:

```
def myfunction:  # pragma: no cover  <- would ignore the whole function
    code
    code
    if something:
        code
        code
    else:  # pragma: no cover <- would ignore the else branch
        code
```

but those are not meant to exclude a whole bunch of code just because it's complicated to write the tests. Then better don't add any pragmas, and try to think of a way to have the code in question tested anyways

**Comment by antespi** [comment]
Well, I don't know how to test a controller, in this case the controller receive a mandrill event, try to figure out for which database it is. Controller reads HTTP Request.

Send method, sents an email, but important code is the one that adds metadata headers  (already tested) 

Process reject event is a rare event, I don't have any real sample (other tests have real ones)

I can add a new test for testing 'Main event process', give me a try.

**Comment by hbrunn** [comment]
@antespi for testing a controller, just instantiate it and call it. or probably the core or server-tools or website projects have invented smarter ways to do this. for the rest, just invent whatever data is necessary.
the tests are not here to give us all a bad time, they are there to avoid having a bad time later

File: mail_mandrill/security/ir.model.access.csv (hunk 1 above)
Comment (antespi) [inline mail_mandrill/security/ir.model.access.csv:5]: This addon is a base addon for getting events from Mandrill. It only depends on mail addon and this is intentional.
There will be more addons that depends on this one and marketing (for instance) that creates security groups and menus in order to give other users access to this info.
BTW I've added read access to base.group_user (Employee), as they have access for reading partners email addresses.

**Comment by antespi** [comment]
Two more tests added and read access to employees

**Comment by antespi** [comment]
@hbrunn I think this addon is ready, I hope anyone review today (OCAReviewDay)

**Comment by rafaelbn** [comment]
Awesome! ;-) :+1:

File: mail_mandrill/controllers/main.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        # Check database selected by mandrill event
+        if not db:
+            _logger.info('%s: No DB selected', info)
+            return False
+        try:
+            registry = openerp.registry(db)
+        except OperationalError:
+            _logger.info("%s: Selected BD '%s' not found", info, db)
+            return False
+        except:
```
Comment (lepistone) [inline mail_mandrill/controllers/main.py:78]: Do we need a catch-all here?

File: mail_mandrill/controllers/main.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        and setting this Post to URL:
+            https://your_odoodomain.com/mandrill/event
+        """
+        if not self._mandrill_validation(**kw):
+            _logger.info('Validation error, ignoring this request')
+            return 'NO_AUTH'
+        events = []
+        try:
+            events = json.loads(kw.get('mandrill_events', '[]'))
+        except:
```
Comment (lepistone) [inline mail_mandrill/controllers/main.py:118]: Here also, maybe we could catch something more specific?

**Comment by lepistone** [comment]
Two minor remarks on exceptions, but this is great! :+1:

File: mail_mandrill/controllers/main.py (hunk 3 above)
Comment (antespi) [inline mail_mandrill/controllers/main.py:118]: Here mandrill must include a `mandrill_events` parameter JSON compliance, so if any exception, no events could be processed. I that case events will be [] and `NO_EVENTS` appears.

What exception do you expect to be catched here?

File: mail_mandrill/controllers/main.py (hunk 2 above)
Comment (antespi) [inline mail_mandrill/controllers/main.py:78]: I don't know if `openerp.registry(db)` could raise more exceptions, this is a high level call. Lots of calls to postgresql libs are done inside, so more exceptions could be raised. In that case, we log connection error and `ALL_EVENTS_FAILED` or `SOME_EVENTS_FAILED` appears.

Comment (lepistone) [inline mail_mandrill/controllers/main.py:78]: In cases where we don't know I have a preference for a crash than catching the unknown. But I see your point, I'm not blocking your PR for that. Thanks! (same on the other comment).

**Comment by rafaelbn** [comment]
@hbrunn are you agree merging?

File: mail_mandrill/README.rst [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/social/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/social/issues/new?body=module:%20mail_mandrill%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+
+
+License
+=======
```
Comment (max3903) [inline mail_mandrill/README.rst]: No need for this section. You have the AGPL badge on top of the README.

Comment (antespi) [inline mail_mandrill/README.rst]: Fixed, thanks

**Comment by max3903** [comment]
:+1:

**Comment by pedrobaeza** [comment]
:+1:

**Pull request merged**